repack.sh

```diff
@@ -1,81 +1,139 @@
 #!/bin/bash
 set -euo pipefail
 
-FILE=""
-OUTPUT=""
+# GLOBALS
+declare -g FILE=""
+declare -g OUTPUT="fwupdates"
+declare -g CAB_ARRAY=()
 
 usage()
 {
-	echo "Usage: $0 [OPTION]..."
+	echo "Usage: $0 <FILE> [OUTPUTDIR]"
 	echo "Repackages Microsoft Surface firmware for fwupd"
 	echo
+	echo "Arguments:"
+	echo "    FILE       The file to repack"
+	echo "                 (can be .msi, .cab, .inf, or a directory)"
+	echo "    OUTPUTDIR  The directory where to save the output"
+	echo "               (Optional; default is '$OUTPUT')"
 	echo "Options:"
-	echo "    -h    This help message"
-	echo "    -f    The file to repack"
-	echo "    -o    The directory where to save the output"
-	exit
+	echo "    -h         This help message"
 }
 
-while getopts ":hf:o:" args; do
-	case "$args" in
-	f)
-		FILE="$OPTARG"
+
+# For backwards compatibility, allow -f and -o flags.
+eval set -- "$(getopt -o 'hf:o:' --long 'help,input:,output:' -- "$@")"
+while true; do
+	case "${1}" in
+	-f|--input)
+		FILE="$2"
+		shift 2
 		;;
-	o)
-		OUTPUT="$OPTARG"
+	-o|--output)
+		OUTPUT="$2"
+		shift 2
 		;;
-	h)
+	-h|--help)
 		usage
+		exit
+		;;
+	--)
+		shift
+		break
 		;;
 	*)
-		echo "ERROR: Invalid command line option '$args'"
-		exit
+		echo "ERROR: Invalid command line option '${1}'"
+		exit 1
 		;;
 	esac
 done
 
+if [ "$FILE" = "" ] && [ $# -gt 0 ]; then
+    FILE="${1}"
+    shift
+fi
+
+if [ "$#" -gt 0 ]; then
+    OUTPUT="${1}"
+    shift
+fi
+
+if [ "$#" -gt 0 ]; then
+    echo "ERROR: Excess arguments: $*"
+    exit 1
+fi
+
 if [ "$FILE" = "" ]; then
 	echo "ERROR: No filename specified!"
-	exit
+	usage
+	exit 1
 fi
 
 if [ "$OUTPUT" = "" ]; then
 	echo "ERROR: No output directory specified!"
-	exit
+	exit 1
 fi
 
-if ! command -v "msiextract" > /dev/null; then
-	echo "ERROR: command 'msiextract' not found, please install the corresponding package"
-	exit
-fi
+for c in msiextract gcab dos2unix; do
+    if ! command -v $c > /dev/null; then
+	echo "ERROR: command '$c' not found, please install the corresponding package"
+	exit 1
+    fi
+done
 
-if ! command -v "gcab" > /dev/null; then
-	echo "ERROR: command 'gcab' not found, please install the corresponding package"
-	exit
-fi
 
-if ! command -v "dos2unix" > /dev/null; then
-	echo "ERROR: command 'dos2unix' not found, please install the corresponding package"
-	exit
-fi
+main()
+{
+    mkdir -p "${OUTPUT}"
+
+    case "${FILE}" in
+	*.msi) repackmsi "${FILE}" "${OUTPUT}"
+	       ;;
+	*.cab) repackcab "${FILE}" "${OUTPUT}"
+	       ;;
+	*.inf) repackinf "${FILE}" "${OUTPUT}"
+	       ;;
+	*) if [ -d "${FILE}" ]; then
+	       repackdir "${FILE}" "${OUTPUT}"
+	   else
+	       echo "==> ${FILE}: Invalid file type!"
+	       exit 1
+	   fi
+    esac
+
+    if [[ ${#CAB_ARRAY[@]} -gt 0 ]]; then
+	echo "Success!"
+	echo "If you wish, you may now install the firmware like so:"
+	echo
+	local f
+	for f in "${CAB_ARRAY[@]}"; do
+	    echo -n "  sudo fwupdmgr install --allow-older --no-reboot-check --force "
```

Comment:
The README has --allow-reinstall but this doesn't, is that intentional?

Reply:
Not intentional. Will fix.